optuna/integration/lightgbm_tuner/optimize.py

```diff
@@ -274,9 +247,15 @@ def __call__(self, trial):
         val_score = self._get_booster_best_score(booster)
         elapsed_secs = t.elapsed_secs()
         average_iteration_time = elapsed_secs / booster.current_iteration()
+
+        if self.model_dir is not None:
+            path = os.path.join(self.model_dir, "{}.pkl".format(trial.number))
+            with open(path, "wb") as fout:
+                pickle.dump(booster, fout)
```

Comment:
(Comment only): I support the use of `pickle.dump` instead of `booster.save_model`.
`pickle.load` may cause problems when the versions of LightGBM in the distributed and development env are different. At least, in my trials, there have been no major problems with the last two versions.
Also, `booster.save_model` does not save many attributes of the booster. It leads to some inconvenience in practical usage.

Reply:
> `booster.save_model` does not save many attributes of the booster. It leads to some inconvenience in practical usage.

Thank you for your comment. I also had a problem when I write the [test case](https://github.com/optuna/optuna/pull/1076/files#diff-20171b34cc0c35c6756d064ab0ccf033R659) because `booster.save_model` does not persist `booster.params`. I think it is the expected behavior of [`lightgbm.Booster`](https://github.com/microsoft/LightGBM/blob/master/python-package/lightgbm/basic.py#L1831).

I think `pickle` is a reasonable choice since the [official example](https://github.com/microsoft/LightGBM/blob/master/examples/python-guide/advanced_example.py) also uses it.